### PR TITLE
Allow dynamic doubleSided and materialThreshold.

### PR DESCRIPTION
**Pull request by prideout**
Note that `materialThreshold` was already dynamic, this just makes it a little easier.

The only way in which the `doubleSided` change could impact existing applications is that materials that explicitly set doubleSided to false will now respect the material's culling mode, rather than forcing it to NONE.

### CODE REVIEW

File: filament/include/filament/MaterialInstance.h [hunk 1]
```diff
@@ -123,6 +123,18 @@ class UTILS_PUBLIC MaterialInstance : public FilamentAPI {
      * @param constant scale factore used to create a constant depth offset for each triangle
      */
     void setPolygonOffset(float scale, float constant) noexcept;
+
+    /**
+     * Overrides the minimum alpha value a fragment must have to not be discarded when the blend
+     * mode is MASKED. Defaults to 0.4 if it has not been set in the parent Material.
+     */
+    void setMaskThreshold(float threshold) noexcept;
```
Comment (bejado) [inline filament/include/filament/MaterialInstance.h]: Update MaterialInstance.java as well?

Comment (prideout) [reply filament/include/filament/MaterialInstance.h]: Good idea, done.  Also updated JS bindings :)

File: filament/src/Material.cpp [hunk 2]
```diff
@@ -204,19 +204,8 @@ FMaterial::FMaterial(FEngine& engine, const Material::Builder& builder)
...
-            // revert culling to default
-            if (mCullingMode == CullingMode::NONE) {
-                mRasterState.culling = CullingMode::BACK;
-            } else {
-                // use the front/back/front&back mode set by the user
-                mRasterState.culling = mCullingMode;
-            }
-        }
+        mDoubleSidedCapability = true;
+        mRasterState.culling = mDoubleSided ? CullingMode::NONE : mCullingMode;
```
Comment (romainguy) [inline filament/src/Material.cpp]: There is a behavior change here. If the culling mode is set to none but double-sided is explicitly disabled, we revert to back face culling in the original code.

File: filament/src/MaterialInstance.cpp [hunk 3]
```diff
@@ -58,6 +61,11 @@ FMaterialInstance::FMaterialInstance(FEngine& engine, FMaterial const* material)
         static_cast<MaterialInstance*>(this)->setParameter(
                 "maskThreshold", material->getMaskThreshold());
     }
+
+    if (material->hasDoubleSidedCapability()) {
+        static_cast<MaterialInstance*>(this)->setParameter(
+                "doubleSided", material->isDoubleSided());
```
Comment (romainguy) [inline filament/src/MaterialInstance.cpp]: We should use a prefix on those built-in parameters. Maybe _ or filament_. Mask threshold should too.

File: libs/filamat/src/shaders/MaterialInfo.h [hunk 4]
```diff
@@ -33,6 +33,7 @@ using CullingMode = filament::backend::CullingMode;
 
 struct UTILS_PUBLIC MaterialInfo {
     bool isLit;
+    bool hasDoubleSidedCapability;
     bool isDoubleSided;
```
Comment (romainguy) [inline libs/filamat/src/shaders/MaterialInfo.h]: Is this still useful?

File: filament/src/Material.cpp (hunk 2 above)
Comment (prideout) [reply filament/src/Material.cpp]: Yes the behavior change is noted in the PR description, this is an intentional change to allow clients to dynamically enable two-sided lighting, which wouldn't be useful if backfaces are culled.

File: libs/filamat/src/shaders/MaterialInfo.h (hunk 4 above)
Comment (prideout) [reply libs/filamat/src/shaders/MaterialInfo.h]: Yes ShaderGenerator needs to know whether or not to emit the `if (doubleSided) {}`.

File: filament/src/MaterialInstance.cpp (hunk 3 above)
Comment (prideout) [reply filament/src/MaterialInstance.cpp]: Done, pushed a new commit for this.

File: libs/filamat/src/shaders/MaterialInfo.h (hunk 4 above)
Comment (romainguy) [reply libs/filamat/src/shaders/MaterialInfo.h]: Isn't `hasDoubleSidedCapability` enough for this? Do we need both `isDoubleSided` and `hasDoubleSidedCapability`?

Comment (prideout) [reply libs/filamat/src/shaders/MaterialInfo.h]: Doh thanks, cleaned this up.

**Pull request merged**